### PR TITLE
[WIP] Index the documents and others links in AI Assist

### PR DESCRIPTION
**Pull request by TwistingTwists**
This makes it way more convenient to use external documentation while using AI Assist. Cursor has a similar feature where `@<url>` is indexed and made part of the AI Assist context.

image

Second example 
image

image

There is further work to make it production ready. Here is how we think about it .

### Make it work
- [x] Run the code
- [x] Index the content of the link
- [x] Identify the link from user messages
- [x] Read the content of the link
- [x] Push the content in the context window before actual OpenAI call 

### Make it right
- [ ] Cache the url content and put in sqlite
- [ ] Make get_url_content async
- [ ] Handle the cases where the content length might be better than the context window
- [ ] Fetching url content might take too long, handle it so 
- [ ] give some indication the url is being index
- [ ] Clean the indexed html content

**or**

- N/A

### CODE REVIEW

File: crates/assistant/Cargo.toml [hunk 1]
```diff
@@ -43,6 +44,9 @@ ui.workspace = true
 util.workspace = true
 uuid.workspace = true
 workspace.workspace = true
+reqwest = { version = "0.11", features = ["blocking"] }
```
Comment (maxdeviant) [inline crates/assistant/Cargo.toml:47]: We already have an HTTP client in Zed that we should be using instead of `reqwest`.

Comment (TwistingTwists) [reply crates/assistant/Cargo.toml:47]: Sure. Noted. Thank you for suggestion.

**Comment by TwistingTwists** [comment]
image
after rebasing with main

**Comment by rgbkrk** [comment]
We definitely want a way to index content!

This fetches the URL content on every message submission, so indexing is a bit of a misnomer here.

**Comment by hargup** [comment]
> This fetches the URL content on every message submission, so indexing is a bit of a misnomer here.

yes, we are aware. This is the first pass and we'll make it right in the next interations :)

**Comment by as-cii** [comment]
Hey @TwistingTwists and @hargup, we really appreciate your work on this. We're currently working on restructuring the assistant quite heavily right now, so I am afraid it would be a bit tricky to accept contributions at the moment, so I am tempted to close this for the time being.

In the new assistant we want to support indexing links, so hopefully we can take inspiration from this pull request to add that behavior.

Thank you and apologies for the delay on this!

**Comment by hargup** [comment]
@as-cii 
Sounds good! Looking forward to the new changes.

**Pull request closed without merging**